pkgs/utils/src/event_loop.zig

```diff
@@ -0,0 +1,529 @@
+const std = @import("std");
+const Thread = std.Thread;
+const Mutex = Thread.Mutex;
+const Allocator = std.mem.Allocator;
+
+const xev = @import("xev");
+
+pub const EventLoopError = error{
+    ShuttingDown,
+};
+
+pub const EventLoop = struct {
+    allocator: Allocator,
+    loop: *xev.Loop,
+    // events from libp2p or other threads will also be pushed on it
+    mutex: Mutex,
+    async_notifier: xev.Async,
+    async_completion: xev.Completion,
+    pending_work: std.ArrayList(WorkItem),
+    stopping: std.atomic.Value(bool),
+
+    pub const WorkItem = struct {
+        callback: *const fn (*anyopaque) anyerror!void,
+        data: *anyopaque,
+    };
+
+    const Self = @This();
+    pub fn init(allocator: Allocator, loop: *xev.Loop) !Self {
+        var async_notifier = try xev.Async.init();
+        errdefer async_notifier.deinit();
+
+        return Self{
+            .allocator = allocator,
+            .loop = loop,
+            .mutex = Mutex{},
+            .async_notifier = async_notifier,
+            .async_completion = undefined,
+            .pending_work = std.ArrayList(WorkItem).init(allocator),
+            .stopping = std.atomic.Value(bool).init(false),
+        };
+    }
+
+    pub fn deinit(self: *Self) void {
+        self.async_notifier.deinit();
+        self.pending_work.deinit();
+    }
+
+    /// Initiates graceful shutdown of the event loop.
+    /// This method signals the event loop to stop, sends a notification for pending work,
+    /// and attempts to process it immediately with a best-effort loop run.
+    /// Should be called before deinit().
+    pub fn stop(self: *Self) void {
```

Comment:
we can add a stop async too, the best practice of xev.loop usage is https://github.com/ghostty-org/ghostty/blob/fb5b8d7968e6d760b53785ba169c751de75ac08d/src/termio/Thread.zig#L240